src/push-manifest.ts

```diff
@@ -0,0 +1,163 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {writeFile} from 'fs';
+import * as path from 'path';
+import {Analyzer, FSUrlLoader} from 'polymer-analyzer';
+
+import {urlFromPath} from './path-transformers';
+import {PolymerProject} from './polymer-project';
+
+
+export interface AddPushManifestOptions {
+  project: PolymerProject;
+  buildRoot: string;
+  bundled?: boolean;
+  path?: string;
+}
+
+// Push Manifest Types Definitions
+export type LinkType = 'document' | 'script' | 'style' | 'image' | 'font';
+const extensionToTypeMapping: {[extname: string]: LinkType} = {
+  '.css': 'style',
+  '.gif': 'image',
+  '.html': 'document',
+  '.png': 'image',
+  '.jpg': 'image',
+  '.js': 'script',
+  '.json': 'script',
+  '.svg': 'image',
+  '.webp': 'image',
+  '.woff': 'font',
+  '.woff2': 'font'
+};
+
+
+export interface PushManifestEntry {
+  [id: string]: {
+    type?: LinkType,
+    weight: 1,
+  };
+}
+export interface PushManifest { [id: string]: PushManifestEntry; }
+
+
+/**
+ * Analyze the given URL and return a unique Set of import URLs analyzed from
+ * that root.
+ */
+async function analyzeFileForImportUrls(
+    analyzer: Analyzer, url: string, ignoreUrls?: Set<string>):
+    Promise<Set<string>> {
+      const analyzedDocument = await analyzer.analyze(url);
+      const analyzedImports = analyzedDocument.getByKind(
+          'import', {externalPackages: true, imported: true});
+      const analyzedImportUrls = new Set();
+      function shouldIgnoreFile(url: string) {
+        return ignoreUrls && ignoreUrls.has(url);
+      }
+
+      for (const analyzedImport of analyzedImports) {
+        if (!shouldIgnoreFile(analyzedImport.url)) {
+          analyzedImportUrls.add(analyzedImport.url);
+        }
+      }
+      return analyzedImportUrls;
+    }
+
+/**
+ * Gets a link type from a file URL. Useful for generating a PushManifestEntry
+ * and needed for the "as=" value on Link headers.
+ */
+function
+getLinkTypeFromUrl(url: string):
```

Comment:
I think we want to track the file type based on how it's referenced and only use the extension as a fallback. If you script src it, it's a script. If you html import it, it's a document. If you link rel=stylesheet it, it's a style.

This is a bit more complicated, because a file could in theory be referenced in multiple different ways, and the service worker shouldn't fall over in that case. I guess this is best effort so last-import-wins should be fine when there's conflicts.